### PR TITLE
Save as image with annotations

### PR DESCRIPTION
**Pull request by arnaud-morvan**
Draw annotations when saving canvas image to file.
Fix http://hub.qgis.org/issues/11503
Fix https://hub.qgis.org/issues/11198
Fix https://hub.qgis.org/issues/5301

### CODE REVIEW

**Comment by nirvn** [comment]
Ahh thanks for fixing this.

BTW, you can add this hub ticket in your commit message to close it: http://hub.qgis.org/issues/11503

File: src/gui/qgsmapcanvas.cpp [hunk 1]
```diff
@@ -800,26 +800,62 @@ void QgsMapCanvas::updateMap()
...
+    painter.begin( &image );
+  }
+
+  // draw annotations
+  QList<QGraphicsItem*> itemList = items();
+  if ( itemList.size() < 1 )
+  {
+    return;
   }
+  QGraphicsItem* item = 0;
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: Use `nullptr` (or skip this entirely, the initial assignment will never be used)

File: src/gui/qgsmapcanvas.cpp [hunk 2]
```diff
@@ -800,26 +800,62 @@ void QgsMapCanvas::updateMap()
...
+
+    painter.save();
+
+    QPointF itemScenePos = item->scenePos();
+    painter.translate( itemScenePos.x(), itemScenePos.y() );
+
+    //a little trick to let the item know that the paint request comes from the canvas
+    item->setData( 1, "canvas" );
+    item->paint( &painter, new QStyleOptionGraphicsItem() );
+    item->setData( 1, "" );
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: It's probably a tiny bit cleaner to use `QVariant()` (invalid) instead of `""` (empty string)

Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: where is this used actually?

File: src/gui/qgsmapcanvas.cpp [hunk 3]
```diff
@@ -800,26 +800,62 @@ void QgsMapCanvas::updateMap()
...
   else //use the map view
   {
-    mMap->contentImage().save( theFileName, theFormat.toLocal8Bit().data() );
+    image = mMap->contentImage().copy();
+    painter.begin( &image );
+  }
+
+  // draw annotations
+  QList<QGraphicsItem*> itemList = items();
+  if ( itemList.size() < 1 )
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: Is it ok/required to bail out before the painter ends and the image is saved?

File: src/gui/qgsmapcanvas.cpp (hunk 1 above)
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: Or... use

    Q_FOREACH( const QGraphicsItem* item, itemListe )

**Comment by m-kuhn** [comment]
I think this looks good in general, just these couple of questions.

You can add a test for extra points :)

**Comment by arnaud-morvan** [comment]
@m-kuhn : Thanks for your comments, I've applied everything except the test (this should be difficult for me now, maybe I could give a try at Girone).

I notice that I need to paint items in reverse order to get the same image as I see in canvas.

I also notice after saving / reopenning project, the z-order is reversed. Maybe I should file a ticket for this.

File: src/gui/qgsmapcanvas.cpp [hunk 4]
```diff
@@ -800,26 +800,55 @@ void QgsMapCanvas::updateMap()
...
+    {
+      continue;
+    }
+
+    painter.save();
+
+    QPointF itemScenePos = item->scenePos();
+    painter.translate( itemScenePos.x(), itemScenePos.y() );
+
+    item->paint( &painter, new QStyleOptionGraphicsItem() );
```
Comment (m-kuhn) [inline src/gui/qgsmapcanvas.cpp]: Is ownership of the option transferred?

Comment (arnaud-morvan) [inline src/gui/qgsmapcanvas.cpp]: Thanks again, I'm noob with c++.
Now that option is on the stack, we should not have memory leak.

**Comment by m-kuhn** [comment]
Code looks good now

> I also notice after saving / reopenning project, the z-order is reversed. Maybe I should file a ticket for this.

This is not a new issue I guess?

**Comment by arnaud-morvan** [comment]
This has nothing to do with this pull request, it is already the case on master.

**Pull request merged**